server/paymentProviders/stripe/virtual-cards.js

```diff
@@ -52,23 +57,111 @@ export const assignCardToCollective = async (cardNumber, expireDate, cvv, collec
   return await models.VirtualCard.create(cardData);
 };
 
+export const processAuthorization = async (stripeAuthorization, stripeEvent) => {
+  const virtualCard = await getVirtualCardForTransaction(stripeAuthorization.card.id);
+
+  if (!virtualCard) {
+    throw new Error(`Virtual card ${stripeAuthorization.card.id} not found`);
+  }
+
+  const host = virtualCard.host;
+
+  await checkStripeEvent(host, stripeEvent);
+
+  const amount = stripeAuthorization.amount;
+  const balance = await host.getBalanceWithBlockedFundsAmount();
+  const connectedAccount = await host.getAccountForPaymentProvider(providerName);
+  const stripe = getStripeClient(host.slug, connectedAccount.token);
+
+  if (balance.value >= amount) {
+    await stripe.issuing.authorizations.approve(stripeAuthorization.id);
+  } else {
+    await stripe.issuing.authorizations.decline(stripeAuthorization.id);
+    throw new Error('Balance not sufficient');
+  }
+
+  const existingExpense = await models.Expense.findOne({
+    where: {
+      VirtualCardId: virtualCard.id,
+      data: { authorizationId: stripeAuthorization.id },
+    },
+  });
+  if (existingExpense) {
+    logger.warn(`Virtual Card authorization already reconciled, ignoring it: ${stripeAuthorization.id}`);
+    return;
+  }
+
+  const vendor = await getOrCreateVendor(
+    stripeAuthorization['merchant_data']['network_id'],
+    stripeAuthorization['merchant_data']['name'],
+  );
+  const hostCurrencyFxRate = await getFxRate('USD', host.currency);
+  const UserId = virtualCard.UserId;
+  const collective = virtualCard.collective;
+  const description = `Virtual Card charge: ${vendor.name}`;
+  const incurredAt = stripeAuthorization.created;
+
+  let expense;
+
+  try {
+    expense = await models.Expense.create({
+      UserId,
+      CollectiveId: collective.id,
+      FromCollectiveId: vendor.id,
+      currency: 'USD',
+      amount,
+      description,
+      VirtualCardId: virtualCard.id,
+      lastEditedById: UserId,
+      status: ExpenseStatus.PROCESSING,
+      type: ExpenseType.CHARGE,
+      incurredAt,
+      data: { authorizationId: stripeAuthorization.id },
+    });
+
+    await models.ExpenseItem.create({
+      ExpenseId: expense.id,
+      incurredAt,
+      CreatedByUserId: UserId,
+      amount,
+    });
+
+    await models.Transaction.createDoubleEntry({
```

Comment:
I don't want us to create the Transaction at this point, we should be waiting for the `issuing_transaction.created` event.